### ISSUE TITLE
Type is not inferred automatically in CE function .Yield parameter when other overloads are present

### ISSUE THREAD

**Issue by Lanayx** [dotnet/fsharp#17837]
I'm building a library that heavily depends on .Yield overloads in CE. I've found the case when it is expected for the type to get inferred automatically, however it is not

**Repro steps**

```fsharp
open System

type Test() =
    member inline _.Yield(value: int -> unit) = ignore
    member inline _.Yield(value: string) = ignore // without this overload issue is not observed
    member inline _.Zero() = ignore

let x =
    Test() {
        yield (fun y -> Console.WriteLine(y))
    }
```

**Expected behavior**

I expect the code above to compile since the `y` type should be inferred to `int`

**Actual behavior**

Compilation error, since `y` requires specifying type explicitly

**Known workarounds**

Specify `y` type explicitly

**Related information**

```
.NET SDK:
 Version:           9.0.100-rc.1.24452.12
 Commit:            81a714c6d3
 Workload version:  9.0.100-manifests.67cd1eb6
 MSBuild version:   17.12.0-preview-24422-09+d17ec720d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-rc.1.24452.12\
 ```

**Comment by vzarytovskii** [comment]
Does it show overloads in the message?

**Comment by vzarytovskii** [comment]
> Does it show overloads in the message?

Just checked, it just checks all candidates, which probably means it's in the generic code or something.

**Comment by Lanayx** [comment]
Right, updated the description with full error message

**Comment by T-Gro** [comment]
I believe this is not a regression, but haven't checked with (even) older versions.

https://sharplab.io/#v2:DYLgZgzgNALiCWwA+B7ADgUwHYAIDKAnhDBgLYCwAUFTAZjgCobEAUAlDgLxU686lkARhgBOOeFmASMOAPoA6AJrwMwACYsAbgENgAVwwhxWGDgC0APhx6s8GB07iA5lhQiMPPgNLCxEqVgyCsqqGjr6hjjEIhJODs6u7jgA9Mk4AO52ABYoeqYwWfAQOCiaosAo2mriEBAGNTiupiiCEKJlap683r7GAUHyAFqiKOxcCW4e1JTAGKYAHlxdjMwwYwDey3w4BCrqOCxgNjvmVgDCKFgQKLPyAOoxJAAy0iwEbGzLAL5AA===

**Comment by T-Gro** [comment]
(btw. I formatted the error message in the issue description to make it more readable and apparent about the overloads)

**Comment by vzarytovskii** [comment]
It's not a regression, no. It's caused by the fact that generics are involved and type is not known by the time WriteLine is called.

**Comment by T-Gro** [comment]
I do not think its by generics. If I replace all the `ignore` (which is what puts in genericity) with `()`, I still get the same issue.

It is a consequence of the order of type inference.
Adding type annotation to `y` fixes it - https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjsNEUAVBHACgEoUBeYlNlPfAIwQCcUAlkmBCEKAPoA6AJoCEwACa0AbgENgAVwQhBSbCgC0APhQakA7IyYoGrdpzw9+QkUjFTZ8pWs3aUOXiEAcysbRgB6cJQAdwsACxgNfWw4gQgUGGU+YBhVBUEICC0ClCQYfRguCD4shTs2ByddV3dJAC0+GAZmMOJiYAR9AA9meqoabG6AbzH2FAw5RRswMxsMECFLQxMAYRgkCBgByQB1QNwAGVFaDHp6MYBfIA=   .


In the general case, I believe this is a method resolution feature proposal for:
- The callsite is known to be passing in a function (not yet fully resolved, but known to be a function)
- There is exactly one overload with a function type argument

**Comment by vzarytovskii** [comment]
> I do not think its by generics. If I replace all the `ignore` (which is what puts in genericity) with `()`, I still get the same issue.
> 
> It is a consequence of the order of type inference. Adding type annotation to `y` fixes it - [sharplab.io#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjsNEUAVBHACgEoUBeYlNlPfAIwQCcUAlkmBCEKAPoA6AJoCEwACa0AbgENgAVwQhBSbCgC0APhQakA7IyYoGrdpzw9+QkUjFTZ8pWs3aUOXiEAcysbRgB6cJQAdwsACxgNfWw4gQgUGGU+YBhVBUEICC0ClCQYfRguCD4shTs2ByddV3dJAC0+GAZmMOJiYAR9AA9meqoabG6AbzH2FAw5RRswMxsMECFLQxMAYRgkCBgByQB1QNwAGVFaDHp6MYBfIA=](https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjsNEUAVBHACgEoUBeYlNlPfAIwQCcUAlkmBCEKAPoA6AJoCEwACa0AbgENgAVwQhBSbCgC0APhQakA7IyYoGrdpzw9+QkUjFTZ8pWs3aUOXiEAcysbRgB6cJQAdwsACxgNfWw4gQgUGGU+YBhVBUEICC0ClCQYfRguCD4shTs2ByddV3dJAC0+GAZmMOJiYAR9AA9meqoabG6AbzH2FAw5RRswMxsMECFLQxMAYRgkCBgByQB1QNwAGVFaDHp6MYBfIA=) .
> 
> In the general case, I believe this is a method resolution feature proposal for:
> 
> * The callsite is known to be passing in a function (not yet fully resolved, but known to be a function)
> * There is exactly one overload with a function type argument

That's what I meant, it's the same issue as https://sharplab.io/#v2:DYLgZgzgPsCmAuACMA7RBeZBXNBPRAtAHyIDKuE8sAtgHQDCA9ihI3LQOoBOAllQDI8UsABS4AlAFgAUEA==, but in case of CE, we can't derive type of function parameter from overload of yield, and it's just `'a` here (which is not generic, but unresolved, yes). Not yet sure how can we changed it.

**Comment by abonie** [comment]
This will need language suggestion to determine how inference should work in such situations

**Comment by Lanayx** [comment]
@abonie I'll be unable to suggest specific changes of how type checking should work and deal with ambiguities, will description like "I'd like typechecking process to look through all possible overloads and try resolve types with those" work?

**Comment by vzarytovskii** [comment]
> @abonie I'll be unable to suggest specific changes of how type checking should work and deal with ambiguities, will description like "I'd like typechecking process to look through all possible overloads and try resolve types with those" work?

Yes, describe when you would like them to work (when they aren't now).

**Comment by T-Gro** [comment]
I think the following change to method resolution would be doable in principle (but still needs the design work to figure out if something else can break because of it)

-The callsite is known to be passing in a function (not yet fully resolved, but known to be a function)
-There is exactly one overload with a function type argument
